src/Stratis.Bitcoin/P2P/Protocol/Behaviors/NetworkPeerBehavior.cs

```diff
@@ -37,16 +50,75 @@ public void Attach(INetworkPeer peer)
             if (this.AttachedPeer != null)
                 throw new InvalidOperationException("Behavior already attached to a peer");
 
-            lock (this.cs)
+            lock (this.lockObject)
             {
                 this.AttachedPeer = peer;
                 if (Disconnected(peer))
                     return;
 
+                this.AttachedPeer.MessageReceived.Register(this.OnMessageReceivedAsync);
+                this.AttachedPeer.StateChanged.Register(this.OnStateChangedAsync);
+
                 this.AttachCore();
             }
         }
 
+        private async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            Type payloadType = message.Message.Payload.GetType();
+
+            foreach (var subscription in this.subscriptions.Where(x => x.Key == payloadType))
+                await subscription.Value(peer, message.Message.Payload, message.Length).ConfigureAwait(false);
+        }
+
+        protected void SubscribeToPayload<T>(OnPayloadReceivedCompact<T> callback) where T : Payload
```

Comment:
You can add a `OnPayloadReceivedCompact` delegate to an ` Dictionary<Type, OnPayloadReceived<object>>`?

Reply:
Yeah, sure. Casting to `object` is probably the only way to keep different generic types in one structure.

Reply:
Ah, I probably misunderstood the question. 
It's not possible because those are different types BUT we have a wrapper: 

```
this.subscriptions.Add(typeof(T), (peer, payload, lenght) => callback(peer, payload as T));
```

So here we actually create a OnPayloadReceived which calls OnPayloadReceivedCompact (which is the same but without the `length` which is just ignored)

And this wrapper OnPayloadReceived is added to the `Dictionary<Type, OnPayloadReceived<object>>`

Reply:
This looks like a good architectural decision to me but if you find it over-engineered or ugly I don't mind having only `OnPayloadReceived` and implementing a `length` input parameter for every callback which will just be ignored in most cases.